runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -10412,227 +10412,590 @@ static TR::Register *inlineIsAssignableFrom(TR::Node *node, TR::CodeGenerator *c
+         break;
+      case TR::Int16:
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, endReg, endReg, endReg);
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, vendReg, vendReg, vendReg);
+         break;
+      case TR::Int32:
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldic, node, endReg, endReg, 2,
+                                         CONSTANT64(0x00000003FFFFFFFC));
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldic, node, vendReg, vendReg, 2,
+                                         CONSTANT64(0x00000003FFFFFFFC));
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+      }
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, valueReg, valueReg, endReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, endReg, valueReg, vendReg);
+
+   // using the serial loop is faster if there are less than 16 items
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi4, node, condReg, vendReg, 16*elementSize);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::blt, node, serialLabel, condReg);
+
+   // load multiplier (anything with more than 4 bytes can be truncated)
+   // 31^0-6 = 1 31 961 29791 923521 28629151 887503681
+   // 31^7 = 667E12CDF           31^8 = C694446F01         31^9 = 180BF449711F 31^10 = 2E97294E4B2C1
+   // 31^11 = 5A44E007B1A55F     31^12 = AEE5720EE830681     31^13 = 152DC8CFCE1DDC99F
+   // 31^14 = 2908B5129F59DB6A41 31^15 = 4F80DED414BE191DDDF 31^16 = 9A09AFBAE83050A9DE01
+   static uint32_t multiplierVectors8[20] = {0x50A9DE01, 0x50A9DE01, 0x50A9DE01, 0x50A9DE01,
+      0xE191DDDF, 0x59DB6A41, 0xE1DDC99F, 0xEE830681, 0x07B1A55F, 0x94E4B2C1, 0xF449711F, 0x94446F01,
+      0x67E12CDF, 887503681, 28629151, 923521, 29791, 961, 31, 1};
+   static uint32_t multiplierVectors16[12] = {0x94446F01, 0x94446F01, 0x94446F01, 0x94446F01,
+      0x67E12CDF, 887503681, 28629151, 923521, 29791, 961, 31, 1};
+   static uint32_t multiplierVectors32[8] = {923521, 923521, 923521, 923521,
+                                               29791, 961, 31, 1};
+   uint32_t *multiplierVector;
+   int mvSize; // multiplierVector's size
+   switch (elementType)
+      {
+      case TR::Int8:
+         multiplierVector = (multiplierVectors8);
+         mvSize = 20;
+         break;
+      case TR::Int16:
+         multiplierVector = (multiplierVectors16);
+         mvSize = 12;
+         break;
+      case TR::Int32:
+         multiplierVector = (multiplierVectors32);
+         mvSize = 8;
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+      }
+
+   // use a similar concept the the TableOfConstants to load the multiplierPtr into the memory
+   // TOC uses relocation data, so we use the same here
+   uint32_t *multiplierPtr = (uint32_t*) fej9->allocateRelocationData(comp, mvSize * sizeof(uint32_t));
```

Comment:
this appears to be a wrong approach, since this can cause gradual memory leak. imagine you have thousands of this call in a java program, you would allocate this constant array thousands of times during JIT-ing. certainly have performance implication too. you still need to set up these arrays statically.